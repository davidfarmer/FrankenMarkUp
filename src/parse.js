/*
The js file for the test interface.

2022.09.27 created

2022.10.17 add global call
2022.10.25 add MathJax support
*/
//import '../spacemath.css'
import {
    remapped_math_tags,
    structural_components,
    level_1_p_peers_containing_p,
    other_level_1_p_peers,
    list_like,
    inlinetags,
    randomtags_containing_p,
    containers,
    display_environments,
    display_subenvironments,
    display_subsubenvironments,
    possibleattributes,
    objects_with_substructure,
    display_math_tags,
    level,
    attribute_like,
    tags_containing_text,
    tags_containing_paragraphs,
    tags_needing_statements,
} from './data.js'
import {splitIntoParagraphs, preprocessAliases, splitAtDelimiters, extract_lists } from './mysplitAtDelimiters.js'
import {reassemblePreTeXt} from './reassemble.js'



// let translateTable = new TranslateTable();

/*
var dictionary;
fetch("dictionary.json").then(
        function(u){ return u.json();}
      ).then(
        function(json){
          dictionary = json;
        }
      )
      */

const PreTeXtDelimiterOf = function(delim) {
    return {left:"<" + delim + ">", right:"</" + delim + ">", tag:delim}
}
const PreTeXtDelimiterOfAttributes = function(delim) {
    return {left:"<" + delim + " ", right:"</" + delim + ">", tag:delim}
}
const LaTeXDelimiterOf = function(delim) {
    return {left:"\\begin{" + delim + "}", right:"\\end{" + delim + "}", tag:delim}
}
export const delimitersFromList = function(lis) {
    if (!Array.isArray(lis)) { return lis }
    let delim_lis = [];
    lis.forEach( (el) => {
        delim_lis.push( PreTeXtDelimiterOfAttributes(el) );
        delim_lis.push( PreTeXtDelimiterOf(el) );
        delim_lis.push( LaTeXDelimiterOf(el) );
    });
    return delim_lis
}

export const PTXdisplayoutput = function(tag) {
    return  { begin_tag: "<" + tag + "",
                       end_tag: "</" + tag + ">",
        before_begin: "\n", after_begin: ">\n",
        before_end: "\n", after_end: "\n"}
}
export const PTXinlineoutput = function(tag) {
    return  { begin_tag: "<" + tag + "",
                       end_tag: "</" + tag + ">",
        before_begin: "", after_begin: ">",
        before_end: "", after_end: ""}
}

export const display_math_delimiters = [
//          {left:"<p>", right:"</p>", tag:"p"},  // for compatibility with PreTeXt!
          {left:"$$", right:"$$", tag:"smen"},
          {left:"\\[", right:"\\]", tag:"smen"},   // these don;t work: not sure why
];
remapped_math_tags.forEach( (el) => {
    display_math_delimiters.push(
        {left:"\\begin{" + el[0] + "}", right:"\\end{" + el[0] + "}", tag:el[1]}
    );
});
display_math_delimiters.push({left: "<md>", right: "</md>", tag: "md"});
display_math_delimiters.push({left: "<me>", right: "</me>", tag: "me"});
display_math_delimiters.push({left: "<mdn", right: "</mdn>", tag: "mdn"});
display_math_delimiters.push({left: "<men", right: "</men>", tag: "men"});

export const paragraph_peer_delimiters = [];

// remapped_tags.forEach( (el) => {
//     paragraph_peer_delimiters.push(
//         {left:"\\begin{" + el[0] + "}", right:"\\end{" + el[0] + "}", tag:el[1]}
//     );
// });


let paragraph_peer_ptx_and_latex_text = [...structural_components, ...level_1_p_peers_containing_p];
let paragraph_peer_ptx_and_latex_text_output = [...paragraph_peer_ptx_and_latex_text, ...list_like];
// plus some tags we don't expect people to type (go back and rethink this)
paragraph_peer_ptx_and_latex_text_output.push("p");
paragraph_peer_ptx_and_latex_text_output.push("statement");


// const paragraph_peer_ptx_and_latex_other = [
//     "figure"
// ];

// Note: no ">" in opening, because could have attributes,
// which are parsed later
paragraph_peer_ptx_and_latex_text.forEach( (el) => {
    paragraph_peer_delimiters.push( PreTeXtDelimiterOfAttributes(el) );
    paragraph_peer_delimiters.push( PreTeXtDelimiterOf(el) );
    paragraph_peer_delimiters.push( LaTeXDelimiterOf(el) );
});
other_level_1_p_peers.forEach( (el) => {
    paragraph_peer_delimiters.push( PreTeXtDelimiterOfAttributes(el) );
    paragraph_peer_delimiters.push( PreTeXtDelimiterOf(el) );
    paragraph_peer_delimiters.push( LaTeXDelimiterOf(el) );
});

let paragraph_peers = Array.from(paragraph_peer_delimiters, ({ tag }) => tag);
paragraph_peers = [...new Set(paragraph_peers)];   //remove duplicates

// console.log("paragraph_peers", paragraph_peers);

let asymmetric_inline_delimiters = [
          {left:"\\(", right:"\\)", tag:"sm"},
//          {left:"|", right:"|", tag:"placeholder"}  // just for testing
];

// need to handle self-closing tags
// also -- for emdash, and abbreviations, i.e., e.g.

inlinetags.forEach( (el) => {
    asymmetric_inline_delimiters.push(  PreTeXtDelimiterOf(el) )
});

/* current;y not used.  See recastSpacedDelimiters
const spacelike_inline_delimiters = [
          {left:"\$", right:"\$", tag:"m"},
          {left:"_", right:"_", tag:"term"},
          {left:"`", right:"`", tag:"c"},
          {left:"'", right:"'", tag:"q"},
          {left:'"', right:'"', tag:"q"},
          {left:"*", right:"*", tag:"em"},
          {left:"**", right:"**", tag:"alert"},
      ];
*/

const do_nothing_markup = {begin_tag: "", end_tag: "",
         before_begin: "", after_begin: "",
         before_end: "", after_end: ""};

export const debugging_output_markup = {begin_tag: "BEGINTAG", end_tag: "ENDTAG",
         before_begin: "BB", after_begin: "AB",
         before_end: "BE", after_end: "AE"};

export const outputtags = {  // start with the quirky ones
    "text" : do_nothing_markup,
    "placeholder" : do_nothing_markup,
    "title": {begin_tag: "<title>", end_tag: "</title>",
         before_begin: "\n", after_begin: "",
         before_end: "", after_end: "\n"},
    };

inlinetags.forEach( (el) => {
    outputtags[el] = { begin_tag: "<" + el + ">", end_tag: "</" + el + ">",
    before_begin: "", after_begin: "",
    before_end: "", after_end: ""}
    });

paragraph_peer_ptx_and_latex_text_output.forEach( (el) => {
    outputtags[el] = PTXdisplayoutput(el)
    });
other_level_1_p_peers.forEach( (el) => {
    outputtags[el] = PTXdisplayoutput(el)
    });
randomtags_containing_p.forEach( (el) => {
    outputtags[el] = PTXdisplayoutput(el)
    });
containers.forEach( (el) => {
    outputtags[el] = PTXdisplayoutput(el)
    });
[...display_environments, ...display_subenvironments, ...display_subsubenvironments].forEach( (el) => {
    outputtags[el] = PTXdisplayoutput(el)
    });

// some special cases
outputtags["ol"] = {begin_tag: "<p>\n<ol>", end_tag: "</ol>\n</p>",
         before_begin: "\n", after_begin: "\n",
         before_end: "\n", after_end: "\n"};
outputtags["ul"] = {begin_tag: "<p>\n<ul>", end_tag: "</ul>\n</p>",
         before_begin: "\n", after_begin: "\n",
         before_end: "\n", after_end: "\n"};

display_math_tags.forEach( (el) => {
    outputtags[el] = {begin_tag: "\n<" + el, end_tag: "</" + el + ">",
         before_begin: "", after_begin: ">\n", // because probably source has the \n
         before_end: "\n", after_end: "\n"};
});

// spacemath
outputtags["sm"] = PTXinlineoutput("m");
outputtags["smen"] = PTXdisplayoutput("men");

outputtags["image"] = {begin_tag: "<image", end_tag: "</image>",  // should not be a special case?
         before_begin: "", after_begin: ">\n", 
         before_end: "\n", after_end: "\n"};
outputtags["description"] = {begin_tag: "<description>", end_tag: "</description>",  // img or image?  should not be a special case?
         before_begin: "\n", after_begin: "",
         before_end: "", after_end: "\n"};

// console.log("in parse.js");

export function fmToPTX(originaltext, wrapper="placeholder"){

//    console.log("fmToPTX", originaltext);
    let originaltextX = preprocessAliases(originaltext);

// console.log("originaltextX", originaltextX);

     // extract title, label, attributes of parent section (currently only title implemented)

      let document_title = "";
      if (originaltextX.match(/^\s*<title>/)) {
          document_title = originaltextX.replace(/^\s*<title>(.*?)<\/title>.*/s,"$1");
          originaltextX = originaltextX.replace(/^\s*<title>(.*?)<\/title>/,"");
      } else if (originaltextX.match(/^\s*\[/)) {
          document_title = originaltextX.replace(/^\s*\[([^\[\]]*)\].*/s,"$1");
          originaltextX = originaltextX.replace(/^\s*\[([^\[\]]*)\]/,"");
      }
   // put latex-style labels on a new line
      let originaltextA = originaltextX.replace(/([^\s])\\label({|\[|\()/g,"$1\n\\label$2");   // }
   // have to preprovess blockquote because (of how we handle attributes) the starting > looks
   // like the end of an opening tag.
      let originaltextB = originaltextA.replace(/\n\s*\n\s*>/g, "\n\n+++sTaRTbQ>");  // preprocess blockquote
      originaltextB = originaltextB.replace(/(\$\$|\\end{equation}|<\/men>|\\end{align}|\\\]) *\n([^\n])/g, "$1\n+++saMePaR$2");  // should take "equation" and "align" from a list

      let originaltextC = originaltextB.replace(/(<diagram)(.*?)(<\/diagram>)/sg, function(x,y,z,w) {
                                  const hiddenz = z.replace(/(<|<\/)definition(>)/g, "$1predefinition$2");
                                  return y + hiddenz + w
                              });
      let findattributes = new RegExp("([^\\n])(\\n *(" + possibleattributes.join("|") + ") *:)", "g");
      originaltextC = originaltextC.replace(findattributes, "$1\n$2");

// end of preprocessor



// console.log("originaltextC", originaltextC);
      // wrap everything in a section
      let tmp1together = {tag: wrapper, content: originaltextC}
      if (document_title) { tmp1together["title"] = document_title }


      let new1 = {...tmp1together};

      const firstdepth = 15;
      for (let depth = 0; depth < firstdepth; ++depth) {
          let trimmed_levels = level.slice(depth);      // need to actually trim them!
          trimmed_levels.forEach( (lev) => {
              new1 = splitAtDelimiters(new1, lev, 0, depth)
              attribute_like.forEach( (attr) => { new1 = extract_lists(new1, attr[0], 0, depth, attr[1]) } );

          } );
      }
//  console.log("preprocessed text 2", new1);
// alert("preprocessed text 2");

      let new7 = {...new1}
// console.log("about to process new7", new7);
// alert("7");
      new7 = splitIntoParagraphs(new7, "all", paragraph_peers);
//   console.log("processed text 7", new7);
//        alert("pause 2");
      let new8 = {...new7}
      new8 = extract_lists(new8, "oneline environments", 0,0, "all");
      new8 = extract_lists(new8, "attributes", 0,0, "all");
//  console.log("processed text 8", new8);
//       alert("pause 3");

      attribute_like.forEach( (attr) => { new8 = extract_lists(new8, attr[0], 0, 0, attr[1]) } );
// console.log("processed text 8b", new8);
//      alert("pause 3");
  // next is maybe overkill, but things like statements contain p's
      new8 = splitIntoParagraphs(new8, "all", paragraph_peers);

      new8 = extract_lists(new8, "blockquotes", 0,0,["p"]);  // meaning: Markdown style

      let new9 = {...new8};

      // why do we extract lists before oneline environments?
      new9 = extract_lists(new9, "extract li", 0,0, "all");   // "all" is wrong, but later code assumes "p"

////////////////////      var tmp1secondsplitPfig = extract_lists(tmp1secondsplitP, "substructure", objects_with_substructure);
//
////////////      var tmp1secondsplitPfigclean = extract_lists(tmp1finalsplit, "clean up substructure", objects_with_substructure);
//  console.log("about to clean up substructure", new9);
//       alert("pause 3");
  // next is maybe overkill, but things like statements contain p's
      new9 = extract_lists(new9, "clean up substructure", 0,0,objects_with_substructure);

      const tmp2 = splitAtDelimiters(new9, asymmetric_inline_delimiters, 0,firstdepth+1, "all", tags_containing_text);

      const tmp3 = splitAtDelimiters(tmp2, "spacelike", 0,firstdepth+1, "all", tags_containing_text);

      //have to do this twice, because of nesting
      const tmp4x = splitAtDelimiters(tmp3, asymmetric_inline_delimiters,0,firstdepth+1, "all", tags_containing_text);
      const tmp4 = splitAtDelimiters(tmp4x, asymmetric_inline_delimiters, 0,firstdepth+1,"all", tags_containing_text);

      const tmp5x = extract_lists(tmp4, "fonts", 0,0,tags_containing_text);
      const tmp5y = extract_lists(tmp5x, "texlike", 0,0,tags_containing_text);

      let tmp5z = splitAtDelimiters(tmp5y, "spacelike", 0,firstdepth+1, "all", tags_containing_text);

      tmp5z = splitAtDelimiters(tmp5z, asymmetric_inline_delimiters,0,firstdepth+1, "all", tags_containing_text);
      tmp5z = splitAtDelimiters(tmp5z, asymmetric_inline_delimiters, 0,firstdepth+1,"all", tags_containing_text);



      const tmp5t = tmp5z;
      const tmp5w = extract_lists(tmp5t, "extract li",0,0, ["p"]);
      const tmp5v = extract_lists(tmp5w, "gather li",0,0, tags_containing_paragraphs);
//  console.log("tmp5v", tmp5v);
//  alert("tmp5v");
      const tmp5u = extract_lists(tmp5v, "absorb math",0,0, tags_containing_paragraphs, "", "", wrapper);
// console.log("tmp5u", tmp5u);
//  alert("tmp5u");
      let tmp5s = extract_lists(tmp5u, "statements",0,0, tags_needing_statements);  // statemetns now part of level
      let tmp5 = extract_lists(tmp5s, "prefigure",0,0, ["prefigure"]);  // statemetns now part of level
 //     let tmp5 = tmp5u;

// console.log("tmp5u == tmp5", JSON.stringify(tmp5u) == JSON.stringify(tmp5));

//       console.log("tmp2 again",tmp2 );
//       console.log("tmp4",tmp4 );
      console.log("tmp5",tmp5 );
      const tmp5p = reassemblePreTeXt(tmp5);

      return tmp5p
};


