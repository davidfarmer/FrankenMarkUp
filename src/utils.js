export const alert = function(...args) {
    //  console.log("alert", ...args);
    if (typeof window !== "undefined" && window.alert) {
        window.alert(...args);
    } else {
        console.log("alert", ...args);
    }
};