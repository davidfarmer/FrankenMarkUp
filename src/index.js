/*
The js file for the test interface.

2022.09.27 created

2022.10.17 add global call
2022.10.25 add MathJax support
*/
import '../spacemath.css'
import {fmToPTX} from './parse.js'

"use strict";
let sourceTextArea = document.getElementById("sourceTextArea");
let echosourceTextArea = document.getElementById("echosourceTextArea");
//let mathmlTextArea = document.getElementById("mathmlTextArea");
//let pretextTextArea = document.getElementById("pretextTextArea");
//let speechTextArea = document.getElementById("speechTextArea");
//let mathJaxArea = document.getElementById("MathJaxArea");

//let translateTable = new TranslateTable();



if (sourceTextArea.addEventListener) {
    sourceTextArea.addEventListener('input', function() {

        const originaltext = sourceTextArea.value;

        let newtext = fmToPTX(originaltext);

        if(echosourceTextArea) {
            echosourceTextArea.innerText = newtext
        }
    }, false);
};

