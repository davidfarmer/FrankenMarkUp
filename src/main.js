import { fmToPTX } from "./parse";

export { fmToPTX as FlexTeXtConvert };