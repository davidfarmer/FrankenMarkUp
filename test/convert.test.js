import { expect, test } from "vitest";
import fs from "fs";
import path from "path";

import { FlexTeXtConvert } from "../src/main";

const mixedDemo = fs.readFileSync(path.join(__dirname, "./mixed-demo.txt"), "utf-8");


// Three examples of how we can use tests in Vitest.
test("demo", () => {
    const converted = FlexTeXtConvert(mixedDemo);
    expect(converted).toMatchSnapshot();
});

test("markdown markup", () => {
    const source = "A _term_ is not to be confused with an *emphasized* word or **alert**."
    const converted = FlexTeXtConvert(source);
    expect(converted).toMatchSnapshot();
});

// Here is an example of a test with input and output explicitly given.  The toMatchSnapshot() is probably better for most tests.
test("math markup", () => {
    const input = "The equation $E=mc^2$ is famous.";
    const output = "\n<p>\nThe equation <m>E = mc^{2}</m> is famous.\n</p>\n";
    expect(FlexTeXtConvert(input)).toBe(output);
});


// Additional tests.
const purePretext = fs.readFileSync(path.join(__dirname, "./pure-pretext.ptx"), "utf-8");
test("pure pretext", () => {
    const converted = FlexTeXtConvert(purePretext);
    console.log(converted);
    expect(converted).toMatchSnapshot();
});