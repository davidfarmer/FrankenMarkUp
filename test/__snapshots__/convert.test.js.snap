// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`demo 1`] = `
"
<section>
<title>FrankenMarkup demo and sample section</title>

<p>
Welcome to a sample of FrankenMarkup: an easy way to write a subset of <pretext/>.
Just mix together <latex/>, Markdown, and <pretext/>, and hope for the best!
</p>

<p>
No need for <c>p</c> tags, just leave a blank line.  Within a paragraph you
can define <term>terminology</term>, use <em>emphasis</em>, or make an <alert>alert</alert>.
You have a choice of math tags:  <m>x^{10} \\lt \\frac{1}{2}</m> or <m>\\sqrt{5x}</m> or even the
official <pretext/> <m>\\sin x \\lt |x| \\ \\ \\text{if}\\ \\ x \\not= 0</m>.  Inside the math tags you can type
either <latex/> or, as in this demo, Space Math.  (Currently the output is <latex/>, which is unfortunate
because information is lost and multiline is not working properly.  It will be better when <pretext/>
allows Space Math or MathML in the source.)
</p>

<note>
<p>
If you start a paragraph with one word followed by a colon,
the word becomes the name of the environment and the
paragraph becomes the contents.
</p>
</note>

<p>
Most block-level components can be typed as <latex/>-style
<c>\\begin{tag}...\\end{tag}</c> or <pretext/>-style \`<tag>...</tag>\`.
If it is only one block of text, just put <c>tag:</c> or <c>Tag:</c> at the
start of that text.
That also works for attributes.
And if it is within a paragraph and is really short and simple,
usually meaning it fits on one line and contains no brackets,
you can type it as \`<tag>...</tag>\`.<fn>Like this!</fn>
</p>

<note>
<p>

You need to mark the beginning and ending if the environment
is more than just one paragraph of text.  Also supported is
markdown lists (- for <c>ul</c> and + for <c>ol</c>) and block quotes, and
(the most common)
<latex/>-style font markup:
</p>

<blockquote>
<p>
 Have you heard the one that starts <q>Möbius and l'Hôpital walk into a bar...</q>?

</p>
</blockquote>
</note>

<p>
These and more features are illustrated below.  To make it realistic,
I started writing an introduction to group theory.
</p>

<note>
<p>
Currently there is no support for sections or other divisions, only the
contents of just one (sub)section.
</p>
</note>

<definition>
<title>Group</title>

<p>

A <term>group</term> is a set <m>G</m> along with a binary operation <m> \\circ </m> such that:
</p>

<p>
<ul>

<li>
The operation is associative: if <m>a,b,c \\in G</m> then <m>a \\circ (b \\circ c) = (a \\circ b) \\circ c</m>.
</li>

<li>
There exists an <term>identity element</term><m>e \\in G</m> such that
<m>e \\circ a = a</m> and <m>a \\circ e = a</m> for all <m>a \\in G</m>.
</li>

<li>
For each <m>a \\in G</m> there exists an <term>inverse</term>
     element <m>b \\in G</m> such that <m>a \\circ b = e</m> and <m>b \\circ a = e</m>.

</li>

</ul>
</p>
</definition>

<p>
To speak of a group one must refer to both the set and the operation.
The sentence <q>The integers are a group.</q> is either incorrect or incomplete,
depending on your perspective.
A correct statement which is similar to that sentence is
<q>The integers are a group under addition.</q>
Any nonempty set can be made into a group by specifying an appropriate
operation <mdash/> don't fall into the trap of assuming the <q>obvious</q>
operation is intended.  Say what you mean, and only use what is (explicitly) said.
</p>

<exercise>
<statement>

<p>
Explain why the condition <em>nonempty</em> was included in the previous paragraph.
</p>

</statement>
</exercise>

<p>
We will now contradict what we wrote about needing to specify the operation!
It is perfectly fine to say <q>Let <m>G</m> be a group.</q>  Because <m>G</m> is a group,
by definition it has an operation, we just haven't
specified a symbol for the operation. We could also write "Let <m>(H,!)</m> be a group."
to indicate that <m>H</m> is the name of the group and <m>!</m> is the symbol for the operation.
Note that in this situation the symbol <q><m>!</m></q> does not mean what it usually means in mathematics.
 If we wish to combine two elements and we don't know the symbol for the operation,
just write them next to each other.
For example: if <m>x</m> and <m>y</m> are elements of <m>G</m>, then so are <m>x y</m> and<nbsp/><m>y x</m>.
<xref ref="oneId"/> uses that notation.
</p>

<p>
As you start your study of group theory,
you may find it helpful to write a symbol for the group operation.
When none has been given, we recommend the symbol <m> \\circ </m> used in our definition.
If you decide to do that, leave a note for the reader,
such as
<q>We will use <m> \\circ </m> for the group operation.</q>
You may have
encountered that symbol when learning about composition of
functions: <m>f \\circ g(x) = f(g(x))</m>.  In that situation, both sides of that equation are
pronounced <q>f of g of x</q>.  That notation for the group operation is meant to be suggestive, for soon
we will see that <em>every</em> group can be thought of as a set of functions,
where the operation is function composition.
</p>

<historical>
<title>Galois invents group theory</title>

<p>

The idea of a group was introduced by the French mathematician Évariste Galois,
who used groups to answer the question:  when is there a nice formula for the
roots of a polynomial?  Most people know the quadratic formula.  There are
formulas for the roots of any cubic (degree<nbsp/>3) or degree<nbsp/>4 polynomial.
It turns out that some higher degree polynomials have nice formulas for their
roots, and some don't.  For example, there is an expression
(complicated, but only using familiar operations)
for the roots of <m>x^{5} - 5 x^{2} - 3</m>, but there is no such formula for the
roots of <m>x^{5} - 5 x^{2} + 3</m>.  Why should those two polynomials
behave so differently?
</p>

<p>
Galois figured out the answer.
He discovered that to each polynomial there can be associated a group,
known today as the <q>Galois group</q> of the polynomial.  Properties of that group,
which we will explore later,
tell you whether there exists a nice formula for the roots.
</p>

<p>
These days <q>Galois theory</q> is an entire branch of research mathematics.
One of the (many) problems which has not been solved is:
Is every finite group the Galois group of a polynomial with rational coefficients?
This is known as <term>The Inverse Galois Problem</term>.
As of this writing, the most recent advance on this problem
was the discovery in November 2024 of a degree 17 polynomial with Galois group<nbsp/><m>\\ \\ \\text{17T7}\\ \\ </m>.
The paper describing their results, which uses methods far beyond basic
group theory, <url href="https://arxiv.org/abs/2411.07857">is available on arXiv</url>,
or you can <url href="https://www.lmfdb.org/NumberField/?galois_group=17T7">see
examples in the LMFDB</url>.

</p>
</historical>

<p>
<ul>

<li>

tags in <c>title</c>
</li>

<li>
<c>figure</c> and bare <c>image</c>
</li>

<li>
<c>foreign</c> and other missing inline tags (such as the new <c>pf</c>)

</li>

<li>

investigate what is missing for a worksheet

</li>

<li>

some Typst compatibility (some of Typst has an
  irreconcilable conflict with other formats).

</li>

<li>
<c>ol</c> and <c>ul</c> as source (not just markdown style)

</li>

</ul>
</p>

<p>
Beyond questions about roots of polynomials,
why are groups important?  In particular, why are those three
axioms natural?
We will answer that question by looking at the details of how to solve the
simplest equations.
</p>

<exploration>
<title>Why the group axioms are natural</title>

<p>

Suppose we want to solve the equation <m>a x = b</m>.  Thinking back to
when you first learned about solving equations, the idea was:
do the same operation on both sides of the equals sign,
and use algebraic identities,
</p>

<p>
First step, multiply on the left by <m>a^{-1}</m>:

<me>
a^{-1} (a x) = a^{-1} b
</me>.
We have just encountered one of the axioms of a group: the existence of
an inverse.
In the above equation we included the parentheses to emphasize that <m>a^{-1}</m> is
multiplying the left side of the equation, which is the <em>quantity</em><m>a x</m>.
Now we regroup (i.e., use associativity) to get

<me>

    (a^-1 a) x = a^{-1} b 
</me>.
What does <m>a^{-1} a</m> equal?  By definition, it equals the identity,
which in the axioms for a group we have denoted<nbsp/><m>e</m>,
so the above equation becomes:

<me>

   e x = a^{-1} b 
</me>.

</p>

<p>
By definition of the identity, <m>e x = x</m>, so we obtain the solution <m>x = a^{-1} b</m>.
</p>

<p>
We see that identity, inverse, and associativity are the
necessary ingredients to solve the simplest equations.
All those properties are necessary,
and
no other properties are needed.
</p>

<p>
The discussion above only used half of the identity axiom and
half of the inverse axiom.
The other halves are needed to solve the equation <m>x a = b</m>.

</p>
</exploration>

<p>
One of the group axioms said <q>there exists</q> an identity.  Does that
mean there can be more than one?  Same question for <q>there exists</q> an inverse.
In the exploration above we talked about <m>a x</m> and also about <m>x a</m>.
Aren't those the same?  Try to keep those questions in mind as we
start looking at some examples of groups.
</p>

<p>
Another issue is the binary operation in the definition of a
group.  In the exploration above there is no explicitly named
operation:  we merely juxtaposed symbols, as in <m>a x</m>, to indicate
that <m>a</m> and <m>x</m> are combined in some way.  What is the operation?
Is it addition? subtraction? multiplication? exponentiation?
The answer is that sometimes there is no answer.  The operation in a group
is just that: an operation <mdash/> a way of combining two objects.
It is not always necessary to give an
explicit meaning to the operation:  just manipulate the
symbols using the axioms.  You will become more
comfortable with this perspective as you gain experience.
</p>

<aside>
<p>
The next material would occur later in the exposition,
probably after some examples.
But this is a demonstration of FrankenMarkup, not a real
textbook on group theory.
</p>
</aside>

<exercise xml:id="ex_hotel mattress">
<title>Flipping a mattress</title>

<statement>

<p>
The Hilbert Hotel has mattresses which are both symmetric and comfortable:
both sides can be used, and you can sleep with your feet pointing either
North or South.  (All the rooms have a window facing either East or West,
and the beds are oriented so that you can easily look at the sunrise or
sunset, if that is your preference.)
</p>

<p>
The manager is concerned that the mattresses are wearing out because
the housekeepers are not rotating them uniformly.  You have received
the following memo:
</p>

<blockquote>
<p>

To help the mattresses last longer, every month we rotate the mattresses.
I suspect that the staff are not keeping track of how they flip the mattresses,
and so parts of the mattress are being used excessively, causing it to
wear out too quickly.
</p>

<p>
Please devise a way to flip the mattress so that the same
operation can be done every time, and over the course of the year all
parts of the mattress will be used equally.  That way it is not
necessary for the staff to keep track of which way it was flipped previously.
It is important that <em>the exact same operation</em> be done every time.

</p>
</blockquote>

<p>
Write two responses to the manager's memo:  a one-sentence
reply which assumes the reader is familiar with the basic concepts of
group theory, and a longer reply which you think the manager might understand.
</p>

</statement>

<hint>
<p>
The fact that a mattress is <m>3</m>-dimensional is irrelevant:
model it as a rectangle.
</p>
</hint>

<hint>
<p>
The manager has assumed that the group of symmetries of a rectangle
can be generated by a single element.  Is that assumption correct?
How many elements are in that group, and what are their orders?
</p>
</hint>
</exercise>

<lemma xml:id="oneId">
<title>The identity element is unique</title>

<statement>

<p>

The identity element of a group is unique.

</p>

</statement>
</lemma>

<proof>
<p>

Suppose <m>e_{1}</m> and <m>e_{2}</m> are identity elements of a group <m>G</m>,
and consider the expression <m>e_{1} e_{2}</m>.
By the first part of the identity axiom we have

<me xml:id="eqn_e_1 e_2 first part">
e_{1} e_{2} = e_{2}
</me>,
because <m>e_{1}</m> is an identity.
Using the second part of the identity axiom we have

<men xml:id="eqn_e_1 e_2 second part">
e_{1} e_{2} = e_{1}
</men>,
because <m>e_{2}</m> is an identity.
Since the left sides of <xref ref="eqn_e_1 e_2 first part"/> and <xref ref="eqn:e_1 e_2 second part"/>
are the same, we see that <m>e_{1} = e_{2}</m>.

</p>
</proof>

<note>
<p>
Labeling and referencing those equations was a bit excessive,
but this is a demo, not a real textbook.
</p>
</note>

<paragraphs>
<title>More samples, for testing</title>

<sidebyside>
<p>

Some <c>inline code</c>. A reference to the hotel exercise: <xref ref="ex_hotel mattress"/>.
<me>
3x^{66} - 8 y = z

</me>

</p>

<p>
This <c>sidebyside</c> will have two parts:  on the left is a paragraph
with text followed by display math, and on the right the paragraph
you are reading now.  Think about why it happens that way.
</p>
</sidebyside>

<blockquote>
<p>
 An egregious block quote in the markdown style.
</p>

<p>
 Checking that multiple paragraph markdown blockquotes are working,
     and that the required markup is not strict.
</p>
</blockquote>

<me>
f(x) = \\begin{cases}
1  \\amp  \\text{ if } x = 1\\\\
0  \\amp  \\text{ if } x \\not= 1\\\\
\\end{cases}

</me>

<mdn>
<mrow>a \\amp b</mrow>
<mrow>b \\amp ct5</mrow>
</mdn>

<p>
<ul>

<li>
apple
</li>

<li>
banana
</li>

<li>
lemon
</li>

</ul>
</p>

<figure>


  This is to check that the contents of a \`figure\` are *not*
  processed as text.  $x^3.14159$  D\\'ej\\\`a vu.

</figure>

<figure xml:id="jiggledX">

<captionThe outcome of one run of <xref ref="p40perturbed"/>.
            </caption>
<image source="jiggledzeros2.png" width="70%">

<description>The unit circle with 40 small randomly placed red squares.
           Inside the circle are thousands of small blue dots.  In the first
quadrant close to the circle the dots generally are in small isolated clusters.
In the other quadrants there are many blue dots clustered near the circle
and a thinner scattering further from the circle.</description>

</image>

</figure>
<image source="FAKE.png" width="70%">

<description>The unit circle with 40 small randomly placed red squares.
           Inside the circle are thousands of small blue dots.  In the first
quadrant close to the circle the dots generally are in small isolated clusters.
In the other quadrants there are many blue dots clustered near the circle
and a thinner scattering further from the circle.</description>

</image>

<p>
Approximate order of next items to implement
(suggestions for additions and rearrangement are welcome):
</p>

<p>
<ul>

<li>
tags in <c>title</c>
</li>

<li>
<c>figure</c> and bare <c>image</c>
</li>

<li>
<c>foreign</c> and other missing inline tags (such as the new <c>pf</c>)
</li>

<li>
investigate what is missing for a worksheet
</li>

<li>
some Typst compatibility (some of Typst has an
  irreconcilable conflict with other formats).
</li>

<li>
<c>ol</c> and <c>ul</c> as source (not just markdown style)
</li>

<li>
more generally: test on pure <pretext/> source (<c>p</c> tags for example)
</li>

<li>
<c>dl</c>
</li>

<li>
investigate what is reasonable for <c>program</c>, <c>listing</c>, and <c>console</c>
</li>

<li>
multiline math
</li>

</ul>
</p>
</paragraphs>

<p>
More things for testing, such as the quadratic formula:
<men>
\\frac{-b \\pm \\sqrt{b^{2} - 4 a c}}{2 a}

</men>

</p>

<p>
And a list:
</p>

<p>
<ol>

<li>
apple
</li>

<li>
banana
</li>

<li>
cherry
</li>

</ol>
</p>

<p>
End of FrankenMarkup demo.
</p>
</section>
"
`;

exports[`markdown markup 1`] = `
"
<p>
A <term>term</term> is not to be confused with an <em>emphasized</em> word or <alert>alert</alert>.
</p>
"
`;

exports[`pure pretext 1`] = `
"

<?xml version="1.0" encoding="UTF-8"?>



<chapter xml:id="pure-pretext" xmlns:xi="http://www.w3.org/2001/XInclude">
    <title>The chapter title</title>


<introduction>

<p>

            This is a sample chapter of pure <pretext/>. It is intended to be used as a test case for the conversion process.

            The content here is written in <pretext/> format, which is a subset of XML designed for educational documents.
        
</p>

</introduction>

<section xml:id="sec-the-first-section">
<title>The first section</title>

<p>

            This is a paragraph that has <term>terms</term>, <em>emphasis</em>, <alert>alerts</alert>, and even some <c>code()</c> in it.
        
</p>

<p>

            In <pretext /> you write math with <tag>m</tag> tags, like this: <m>\\sin(x) = \\cos(x + \\frac{\\pi}{2})</m>.  You can also write an math equation like this:
            
<me>

                f(x) = \\begin{cases}
                    x^2 \\amp \\text{if } x \\geq 0 \\\\
                    -x^2 \\amp \\text{if } x \\lt 0
                \\end{cases}
            
</me>

</p>
</section>


</chapter>
"
`;
